### PR TITLE
Fix #8 - Migration script to backup database file and restart the s…

### LINKED ISSUE #8: Issue database needs a migration script (for webcompat - issues.db backup)
More details about the script here : 
https://github.com/webcompat/webcompat.com/issues/968
https://github.com/webcompat/webcompat.com/issues/1131

### PR DESCRIPTION
**Pull request by deepthivenkat**
…erver

r? @miketaylr @karlcow

### CODE REVIEW

File: webcompat_migration_backup_db.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    assert repo.refs['origin/master'] == repo.remotes.origin.refs.master
+    origin = repo.remotes.origin
+    index = repo.index
+    db_path = path + '/backup_db'
+    if os.path.exists(path + '/issues.db'):
+        if os.path.exists(db_path) and os.path.isdir(db_path):
+            #           check if a previous version of a backup exist and delete it
+            print 'Backup database already exists. Removing older version of the back up'
+            call('rm -r backup_db', shell=True)
+            call('git rm -r backup_db/.', shell=True)
```
Comment (karlcow) [inline webcompat_migration_backup_db.py:27]: `git rm` is used only for things which are committed in the tree.
We discussed about it https://github.com/webcompat/webcompat.com/issues/1131#issuecomment-234347508 and decided that we would not commit it. And just create `.gitignore` so it's not becoming accidently committed.

File: webcompat_migration_backup_db.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    assert not repo.bare
+    assert repo.refs['origin/master'] == repo.remotes.origin.refs.master
+    origin = repo.remotes.origin
+    index = repo.index
+    db_path = path + '/backup_db'
+    if os.path.exists(path + '/issues.db'):
+        if os.path.exists(db_path) and os.path.isdir(db_path):
+            #           check if a previous version of a backup exist and delete it
+            print 'Backup database already exists. Removing older version of the back up'
+            call('rm -r backup_db', shell=True)
```
Comment (karlcow) [inline webcompat_migration_backup_db.py:26]: You are already importing the `os` module, so it seems not necessary here to import the subprocess module. You could directly use the [`os.remove`](https://docs.python.org/2/library/os.html#os.remove) for a file or [`shutil.rmtree`](https://docs.python.org/2/library/shutil.html#shutil.rmtree) for non-empty directory and its content. (there is `os.rmdir` for empty directories.)

File: webcompat_migration_backup_db.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            call('git rm -r backup_db/.', shell=True)
+        print 'Switching to branch : master'
+        call('mkdir backup_db', shell=True)
```
Comment (karlcow) [inline webcompat_migration_backup_db.py:29]: recreating the directory might be overkill, when you could only delete the file inside the directory.

File: webcompat_migration_backup_db.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        try:
+            repo.heads.master.checkout()
+        except GitCommandError:
+            print 'Please make sure your current branch is clean and all the changes are either committed or stashed' \
+                  'before switching to master branch'
+
+        call('mv issues.db backup_db/' + issue_backup_db, shell=True)
+        index.add([path + '/backup_db/' + issue_backup_db])
+        index.commit('Adding new backup database at ' + time_stamp)
```
Comment (karlcow) [inline webcompat_migration_backup_db.py:38]: We do not want to do that.
Discussed in https://github.com/webcompat/webcompat.com/issues/1131#issuecomment-234347508

File: webcompat_migration_backup_db.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        # push the backup database file to master branch of the repository
+        origin.push()
+        call('python extract_id_title_url.py 1', shell=True)
+        call('python run.py&', shell=True)
```
Comment (karlcow) [inline webcompat_migration_backup_db.py:42]: This should probably be handled by the deployment script. If I remember @miketaylr is having a fabric script or something.

File: webcompat_migration_backup_db.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        except GitCommandError:
+            print 'Please make sure your current branch is clean and all the changes are either committed or stashed' \
+                  'before switching to master branch'
+
+        call('mv issues.db backup_db/' + issue_backup_db, shell=True)
+        index.add([path + '/backup_db/' + issue_backup_db])
+        index.commit('Adding new backup database at ' + time_stamp)
+        # push the backup database file to master branch of the repository
+        origin.push()
+        call('python extract_id_title_url.py 1', shell=True)
```
Comment (karlcow) [inline webcompat_migration_backup_db.py:41]: Ah here is the current call for the DB dump.

Comment (karlcow) [inline webcompat_migration_backup_db.py:41]: The issue is that this script is not currently doing a full dump of the DB. @deepthivenkat do you plan on working on this after? :)

Comment (deepthivenkat) [inline webcompat_migration_backup_db.py:41]: I have added a new commit for the  script to extract the full dump  :)

File: webcompat_migration_backup_db.py (hunk 5 above)
Comment (deepthivenkat) [inline webcompat_migration_backup_db.py:42]: I need to run python run.py / the deployment script to regenerate issues.db with the new/modified schema.

Before calling `python dump_webcompat_to_db.py` I need the new issues.db file.

python run.py is for localhost testing. Should I remove this and add the fabric script here?

File: dump_webcompat_to_db.py [hunk 7]
```diff
@@ -47,7 +60,7 @@ def main():
     # stuff data into database..
     for bug in data['bugs']:
         db_session.add(
-            Issue(bug['id'], bug['summary'], bug['url'], bug['body']))
+            Issue(bug['id'], bug['summary'], bug['url'], extract_domain_name(bug['url']), bug['body'], bug['state'], bug['creation_time'], bug['last_change_time']))
```
Comment (miketaylr) [inline dump_webcompat_to_db.py:63]: Can you run pep8 on this file? This doesn't follow webcompat style guidelines.

https://github.com/webcompat/webcompat.com/blob/master/CONTRIBUTING.md#python

File: webcompat_migration_backup_db.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/python
+import datetime
+import time
+from subprocess import call
+
+import os
+from git import Repo, GitCommandError
```
Comment (miketaylr) [inline webcompat_migration_backup_db.py:7]: Where does this module come from? I don't see it in requirements.txt

File: webcompat_migration_backup_db.py (hunk 2 above)
Comment (miketaylr) [inline webcompat_migration_backup_db.py:26]: Yes, shelling out to `rm -r` seems scary here.

File: webcompat_migration_backup_db.py (hunk 5 above)
Comment (miketaylr) [inline webcompat_migration_backup_db.py:42]: Why are you using shell commands to call python modules?

I guess I'm not 100% sure why this lives outside the webcompat.com repo either.

Comment (miketaylr) [inline webcompat_migration_backup_db.py:42]: > Should I remove this and add the fabric script here?

The fabfile is not going to be added to a public repo -- it has deployment secrets.

File: webcompat_migration_backup_db.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        call('python dump_webcompat_to_db.py 1', shell=True)
+    else:
+        raise IOError('FileNotFound : No database file found to backup. Please run python run.py to create the db file')
+
+if __name__ == '__main__':
+    WATCHED_FILES = [path + '/issues.db']
+    WATCHED_FILES_MTIMES = [(f, time.ctime(os.path.getmtime(f))) for f in WATCHED_FILES]
+    #   Read the last modified time property of the issues.db file. If there is any change, run the migration script
+    while True:
+        for f, mtime in WATCHED_FILES_MTIMES:
```
Comment (miketaylr) [inline webcompat_migration_backup_db.py:53]: Can you explain the logic here? Is the idea that this script is a long-running process that does a backup whenever the db is updated? Isn't the db updated every time an issue is opened?

**Comment by miketaylr** [comment]
@deepthivenkat can you summarize the strategy and design here, before addressing review comments? Like, what is the main entry point, by who and when is this script run, what are the expected outcomes.

Right now it just looks like it copies issues.db and then puts fresh data into a new issues.db. Why are we doing this? Presumably something to do with migrating db schemas -- but I don't see any code to handle the actual schema changes.

I'm also concerned about the idea of creating and populating a new db between killing and starting the app. What if GitHub is down, or slow? Or what happens when there are 10,000 or 100,000 issues? Do we hit API limits? Does this take 2 or 10 minutes, and does that mean the site is down until it's finished?

**Comment by miketaylr** [comment]
I also have concerns generally about basing this on extract_id_title_url.py -- it was written by Hallvord for a different project. RIght now it's not up to date (for example, it doesn't know about `needstriage` and it only collects info about Firefox related bugs. 

Have you verified that the database created from the `extract_data` method matches what's on GitHub?

File: webcompat_migration_backup_db.py (hunk 4 above)
Comment (deepthivenkat) [inline webcompat_migration_backup_db.py:38]: I think i proceeded with the script without properly understanding [webcompat/webcompat.com#1131 (comment)](https://github.com/webcompat/webcompat.com/issues/1131#issuecomment-234347508)

Should I add the dump of the db in [webcompat.com project root folder](https://github.com/webcompat/webcompat.com) inside a folder that is git ignored?

or should i use an open source backup software like [bacula](http://blog.bacula.org/what-is-bacula/) since the dump becomes larger everyday?

**Comment by deepthivenkat** [comment]
@miketaylr As you mentioned in IRC, few lines in extract_id_title_url.py like 

https://github.com/webcompat/issue_parser/blob/master/extract_id_title_url.py#L84 

is currently unused. Should there be another script for extracting issues specifically for webcompat inside webcompat.com root folder?

**Comment by deepthivenkat** [comment]
Summarising the strategy:

The script would be run by the person who is handling the fabfile for deployment. The script will update the db schema, and create new data dump for the schema, backup old db inside a .gitignored folder in webcompat root repository. 

The entry point would be to use one of the following tools to handle 

1) [SQLAlchemy](https://sqlalchemy-migrate.googlecode.com/files/sqlalchemy-migrate-doc-0.7.2.pdf) 
2) [Alembic](http://alembic.zzzcomputing.com/en/latest/tutorial.html#the-migration-environment) to handle the schema changes.

If I choose the second option, I will be able to use [flask-migrate](http://blog.miguelgrinberg.com/post/flask-migrate-alembic-database-migration-wrapper-for-flask) that runs on alembic.

The github api rate limits is 5000 per hour for authenticated users and it is linked to the github account. The number of issues we have currently is already about half of maximum limit. So if we run extract_id_title_url.py a couple of times, we will end up with 403 status code.

So it is not very viable to kill the app while populating the new db.

Suggestions: 

We can start the app as soon as running the script for db schema migration(yet to be done using flask-migrate) is complete.

When make the db schema changes, the db will be locked for changes. Once the schema change is done, the webhooks we have in [webhooks/**init**.py](https://github.com/webcompat/webcompat.com/blob/master/webcompat/webhooks/__init__.py) may try to insert the newly opened issue which is ok.

Solutions:

1) Extract the issues from github before killing the app and handle only dump_to_db after killing the app.

Suggestions for extracting issues:

1) We can make use of [Github-Backup](https://github.com/joeyh/github-backup). I read the following from their documentation:

> Bear in mind that this uses the GitHub API; don't run it every 5 minutes. GitHub rate limits the API to some small number of requests per hour when used without authentication. To avoid this limit, you can set GITHUB_USER and GITHUB_PASSWORD (or GITHUB_OAUTH_TOKEN obtained from https://github.com/settings/tokens) in the environment and it will log in when making (most) API requests.
> 
> Anyway, github-backup does do an incremental backup, picking up where it left off, so will complete the backup eventually even if it's rate limited.

2) Write a new extract_dump_issues python script to retrieve issues and dump to db as and when the issues are retrieved.. We can rely on 

X-RateLimit-Limit: 5000
X-RateLimit-Remaining: 4987
X-RateLimit-Reset: 1350085394 

and add sleep until the X-RateLimit-Reset and retrieve issues again.

If the [webhooks/**init**.py](https://github.com/webcompat/webcompat.com/blob/master/webcompat/webhooks/__init__.py)
tries to enter a newly created issues while the db dump is happening in the extract_dump_issues python script, the db will be locked for editing which will throw:

`sqlalchemy.exc.OperationalError: (OperationalError) database is locked`

We can edit the webhook code to handle this error and attempt dumping the data after a time period.

We can also move this discussion and files to webcompat repository and let issue_parser to be in peace ^_^ 

@miketaylr @karlcow

**Comment by miketaylr** [comment]
> The script will update the db schema

This current PR doesn't touch any db schema though, right? (Or am I missing something)?

> backup old db inside a .gitignored folder in webcompat root repository

I think we should start here. Doing a "back-up" is simple, it's literally copying a file and moving it somewhere. But we also want some mechanism to bootstrap a new database with current GitHub information.

Right now, the issues.db is 100% unused (and doesn't even contain 100% of issues). If we made any schema changes today, and re-built the DB nothing would break. So I think it's probably too early to add the complexity of alembic, etc when we don't need it yet.

> 2) Write a new extract_dump_issues python script to retrieve issues and dump to db as and when the issues are retrieved

I think this is what we should do first. It should live in the webcompat.com repo as well -- this issue_parser repo is a side-project that was not written with our needs in mind. github-backup
can be used for inspiration, but it shouldn't be too complicated to just query the GitHub API and pull the data we want.
